ttljob: avoid using different versions of enum; use a desc.Collection tied to the correct txn

rafiss opened the pull request:
### ttl: use a desc.Collection tied to the txn for TTL job metadata

Previously, we would pass in a desc.Collection from the flowCtx while gathering metadata for the TTL job, such as column names and types.

This was incorrect, since it meant that the lease on any descriptors fetched in those transactions would never be released until the job completed.

###  ttljob: make an error message slightly more helpful

The new message assures the user that the job will run again.

### ttljob: avoid using different versions of enum

This patch changes the SELECT/DELETE queries so the use the logical string
representation of enums, rather than the encoded datum. This avoids an
issue that was caused by the enum descriptor's version being passed in
as part of the query placeholder argument.

fixes https://github.com/cockroachdb/cockroach/issues/145908
Informs: https://github.com/cockroachlabs/support/issues/3246

Release note (bug fix): Fixed a bug that could cause the row level TTL
job to fail with a "comparison of two different versions of enum" error
if an ENUM type referenced by the table experienced a schema change.

rafiss committed:
ttl: use a desc.Collection tied to the txn for TTL job metadata

Previously, we would pass in a desc.Collection from the flowCtx while
gathering metadata for the TTL job, such as column names and types.

This was incorrect, since it meant that the lease on any descriptors
fetched in those transactions would never be released until the job
completed.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145374)
<!-- Reviewable:end -->

spilchen reviewed (approved):
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145374)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_plans_test.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145374#-OP0cRid8pqUfp3pzMb6:-OP0cRid8pqUfp3pzMb7:b-iim34h) ([raw file](https://github.com/cockroachdb/cockroach/blob/5be839777ce4ff9a53513c7bc6ce7e07ecd28a31/pkg/sql/ttl/ttljob/ttljob_plans_test.go#L49)):*
> ```Go
> 	require.NoError(t, err)
> 
> 	execCfg := s.ExecutorConfig().(sql.ExecutorConfig)
> ```

nit: I recently added a TTL test (TestRetryDeleteBatch in ttljob_processor_internal_test.go) that had to set the flowCtx.Descriptors. It's worth checking if that's still necessary after this change.


<!-- Sent from Reviewable.io -->

rafiss commented:
Thanks! I'll look at that test.

I'll also add a test case here to prove that my change had an effect -- specifically, schema changes during the TTL job should no longer be blocked.

rafiss committed:
sql: fix typo in TTL test

Release note: None

rafiss committed:
tree: return error instead of panicking

This avoids the risk of a faulty enum comparison causing a node to
crash.

Release note: None

rafiss committed:
ttljob: make an error message slightly more helpful

The new message assures the user that the job will run again.

Release note: None

rafiss commented:
This is ready for another review. I was able to write a test that shows that schema changes can succeed now, and I also was able to repro the error we are seeing in https://github.com/cockroachlabs/support/issues/3246

spilchen reviewed (approved):
Good test. I'm glad we have a reliable repro for the enum compare issue. I just had a couple of nits.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r2, 6 of 6 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145374#-:-OP5eGks5kL10-34ggR-:bw6m007)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_test.go` line 414 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145374#-OP5bJw30dWi57raZPDk:-OP5bJw30dWi57raZPDl:b-izl03x) ([raw file](https://github.com/cockroachdb/cockroach/blob/192314f9ce0b34972bcba98147bf025a184dfd5e/pkg/sql/ttl/ttljob/ttljob_test.go#L414)):*
> ```Go
> 	// Since the enum type is used in the primary key and we just modified it,
> 	// the job will fail.
> 	th.waitForScheduledJob(t, jobs.StateFailed, "comparison of two different versions of enum USER DEFINED ENUM: public.typ oid")
> ```

nit: this error is an assertion failure, so I assume it's temporary until the underlying issue is resolved. Should we add a TODO comment to clarify this?
___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 97 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145374#-OP5cwh19lEiEtfSL-am:-OP5cwh19lEiEtfSL-an:bpxw2jb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a06dea22438e9f821f536b703969344a480bcce4/pkg/sql/ttl/ttljob/ttljob.go#L97)):*
> ```Go
> 			return pgerror.Newf(
> 				pgcode.ObjectNotInPrerequisiteState,
> 				"found a recent schema change on the table at %s, aborting and waiting for the next scheduled job",
> ```

nit: I agree that we should indicate the job will run again. But using the term _waiting_ suggests a synchronous process will wait for this to happen. So, I think it could still be confusing. Do we just retry the job immediately? Some other suggestions:
- found a recent schema change on the table at %s; aborting—job will retry shortly.
- found a recent schema change on the table at %s; aborting this run, which will be retried automatically.
___
*[`pkg/sql/schema_changer_test.go` line 6914 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145374#-OP5XJ187fu6mZPsACsX:-OP5XJ187fu6mZPsACsY:b-ddtk6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/592efbc5365cdd13f8ecd7156ec743b539cb44ad/pkg/sql/schema_changer_test.go#L6914)):*
> ```Go
> 			setup: createNonTTLTable,
> 			schemaChange: `
> ALTER TABLE t.test SET (ttl_job_cron = '@daily');
> ```

Good catch. I guess this wasn't caught before because the schema change was expected to fail anyway.


<!-- Sent from Reviewable.io -->

rafiss committed:
ttljob: avoid using different versions of enum

This patch changes the SELECT/DELETE queries so they use the logical string
representation of enums, rather than the encoded datum. This avoids an
issue that was caused by the enum descriptor's version being passed in
as part of the query placeholder argument.

Release note (bug fix): Fixed a bug that could cause the row level TTL
job to fail with a "comparison of two different versions of enum" error
if an ENUM type referenced by the table experienced a schema change.

spilchen reviewed (commented):
> Reviewed 2 of 2 files at r6, 8 of 8 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145374)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_query_builder.go` line 156 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/145374#-OP6jZuXEr7-BCJMvdAx:-OP6jZuXEr7-BCJMvdAy:bzbvg4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7253bfaea67b72cfa582c5325e5e6e2cc625e3a/pkg/sql/ttl/ttljob/ttljob_query_builder.go#L156)):*
> ```Go
> 	for i, arg := range b.cachedArgs {
> 		if enum, ok := arg.(*tree.DEnum); ok {
> 			b.cachedArgs[i] = enum.LogicalRep
> ```

Should we duplicate this logic a few lines down where we update `b.cachedArgs` for the last fetched row? Since `deleteQueryBuilder` uses `cachedArgs`, and the cached arguments seem to be derived from the fetched rows, should we apply this logic to `DELETE` operations as well? Or is this behavior only relevant for `SELECT`?
___
*[`pkg/sql/ttl/ttljob/ttljob_test.go` line 388 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/145374#-OP6haBR7vcWoovFdfa6:-OP6haBR7vcWoovFdfa7:b2k2gk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7253bfaea67b72cfa582c5325e5e6e2cc625e3a/pkg/sql/ttl/ttljob/ttljob_test.go#L388)):*
> ```Go
> 	// Prior to https://github.com/cockroachdb/cockroach/pull/145374, the job
> 	// would fail with a "comparison of two different versions of enum" error.
> 	th.waitForScheduledJob(t, jobs.StateSucceeded, "comparison of two different versions of enum USER DEFINED ENUM: public.typ oid")
> ```

Do we need the error regex if we expect success now?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145374)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob_query_builder.go` line 156 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/145374#-OP6jZuXEr7-BCJMvdAx:-OPBpeR60pOrnDMyLuZa:b-naj446) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7253bfaea67b72cfa582c5325e5e6e2cc625e3a/pkg/sql/ttl/ttljob/ttljob_query_builder.go#L156)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Should we duplicate this logic a few lines down where we update `b.cachedArgs` for the last fetched row? Since `deleteQueryBuilder` uses `cachedArgs`, and the cached arguments seem to be derived from the fetched rows, should we apply this logic to `DELETE` operations as well? Or is this behavior only relevant for `SELECT`?
</blockquote></details>

the `SELECT` uses AS OF SYSTEM TIME so it's easier for it to hit the error, but i think `DELETE` can too. i'll make the change.
___
*[`pkg/sql/ttl/ttljob/ttljob_test.go` line 388 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/145374#-OP6haBR7vcWoovFdfa6:-OPBpbBt3fw4cYwwx6JQ:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7253bfaea67b72cfa582c5325e5e6e2cc625e3a/pkg/sql/ttl/ttljob/ttljob_test.go#L388)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Do we need the error regex if we expect success now?
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> Reviewed 8 of 8 files at r8, 10 of 10 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145374)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss)



<!-- Sent from Reviewable.io -->
